Back out "[Inductor] Fallback scatter when src dtype is bf16 (#113204)"

nmacchioni opened the pull request:
Summary: Revert due to Llama 7b performance regression on Mi250x (83tok/s -> 79.5tok/s, ~4% regression)

Test Plan: CI

Differential Revision: D51287379




cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler

nmacchioni commented:
Reverting #113204

nmacchioni commented:
@oulgen

nmacchioni commented:
@pytorchbot label "topic: not user facing"

oulgen commented:
I'm on PTO so I'll lean on @Chillee and @eellison here. We can revert first and discover what happened later. Are we confident that this PR is the cause since tl.atomic_add does not work at all with bf16.

nmacchioni commented:
> I'm on PTO so I'll lean on @Chillee and @eellison here. We can revert first and discover what happened later but just to be sure, are we confident that this PR is the cause since `tl.atomic_add` does not work at all with bf16. I don't see any data on github or phabricator.

Yes, I'm sure this is the cause of the regression. My guess is that this fallback is not required for AMD GPUs.

oulgen reviewed (approved):
Please share instructions to measure so that I can check before relanding

nmacchioni commented:
> Please share instructions to measure so that I can check before relanding

I'll ping you internally.

nmacchioni committed:
Back out "[Inductor] Fallback scatter when src dtype is bf16 (#113204)" (#113599)

Summary:

Revert due to Llama 7b performance regression on Mi250x (83tok/s -> 79.5tok/s, ~4% regression)

Test Plan: CI

Reviewed By: xw285cornell

Differential Revision: D51287379

eellison commented:
Can you just skip the fallback if the cuda device is amd then ?

nmacchioni commented:
Unfortunately I have no bandwidth whatsoever to forward fix (I know it would be very simple but I just don't have the time to verify perf numbers are ok on amd/nvidia) and I need this to land by eod so will stick to a revert.

oulgen commented:
> Can you just skip the fallback if the cuda device is amd then ?

Do you know how to check whether the cuda device is amd or NVIDIA? Also when I restricted this to device == cuda, some tests failed. So I assume fallback is required for more than just NVIDIA

eellison commented:
@oulgen : https://github.com/pytorch/pytorch/blob/main/torch/_inductor/graph.py#L264

eellison commented:
Since this didn't land yesterday - can we make the amd change ? we dont usually do correctness reverts for smallish perf regressions.

oulgen commented:
I can put a pr up

On Tue, Nov 14, 2023 at 7:11 AM eellison ***@***.***> wrote:

> Since this didn't land yesterday - can we make the amd change ? we dont
> usually do correctness reverts for smallish perf regressions.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/pytorch/pytorch/pull/113599#issuecomment-1810717194>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AA6SNOVRAGDLINUIXEGEIQTYEOQ4FAVCNFSM6AAAAAA7J7MNNKVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMYTQMJQG4YTOMJZGQ>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

oulgen commented:
https://github.com/pytorch/pytorch/pull/113655

eellison commented:
No longer going with this pr - closing.
